nussl/core/mixing.py

```diff
@@ -3,11 +3,139 @@
 AudioSignal objects. 
 """
 import copy
-
 import numpy as np
-
 from . import AudioSignal
+from .constants import MIN_LOUDNESS, MAX_LOUDNESS
+import scipy
+import warnings
+
+def convolve(signal, other, method='auto', normalize=True, 
+             scale=True):
+    """
+    Convolves signal one with signal two. There are three
+    cases:
+    
+    1. s1 is multichannel and s2 is mono.
+    -> s1's channels will all be convolved with s2.
+    2. s1 is mono and s2 is multichannel.
+    -> s1 will be convolved with each channel of s2.
+    3. s1 and s2 are both multichannel.
+    -> each channel will be convolved with the matching 
+        channel. If they don't have the same number of
+        channels, an error will be thrown.
+
+    Args:
+        other: AudioSignal with which to convolve the two signals.
+        method: A string indicating which method to use to calculate the convolution. Options are:
+            direct: The convolution is determined directly from sums, the definition of convolution.
+            fft: The Fourier Transform is used to perform the convolution by calling fftconvolve.
+            auto: Automatically chooses direct or Fourier method based on an estimate of which is faster (default).
+        normalize: Whether to apply a normalization factor which will prevent clipping. Defaults to True.
+        scale: Whether to scale the output convolved signal to have the same max as the input, so they are of roughly equal loudness.
+    
+    Example:
+
+    >>> import nussl
+    >>> path = nussl.efz_utils.download_audio_file('schoolboy_fascination_excerpt.wav')
+    >>> sig = nussl.AudioSignal(path)
+    >>> path = nussl.efz_utils.download_audio_file('h072_Bar_2txts.wav')
+    >>> ir = nussl.AudioSignal(path)
+    >>> convolved = nussl.mixing.convolve(sig, ir)
+    >>> # Or apply with fluent interface
+    >>> convolved = sig.convolve(ir) 
+    
+    """
+    output = []
+    factor = 1.0
+
+    signal = copy.deepcopy(signal)
+    other = copy.deepcopy(other)
+
+    if normalize: factor = np.sum(np.abs(other.audio_data) ** 2)
+    
+    if signal.num_channels != 1 and other.num_channels != 1:
+        if signal.num_channels != other.num_channels:
+            raise RuntimeError(
+                "If both signals are multichannel, they must have the " 
+                "same number of channels!")
+        for s1_ch, s2_ch in zip(signal.get_channels(), other.get_channels()):
+            convolved_ch = scipy.signal.convolve(
+                s1_ch, s2_ch / factor, mode='full', method=method)
+            output.append(convolved_ch)
+    else:
```

Comment:
Pedantic: Maybe add a comment saying that at least one of these for loops is over 1 item. Wasn't clear to me at first.